### PR TITLE
Fixed issue #9656

### PR DESCRIPTION
**Pull request by SamMousa**
Refactored BaseActiveRecord to not use $_attributes directly but instead use getAttribute, setAttribute and hasAttribute.

Passes all but 1 test in the mysql group. (The test wrongly assumes that NULL + 1 = 1 in MySQL, this is unrelated to my changes)

### CODE REVIEW

**Comment by dynasource** [comment]
looks a lot cleaner

**Comment by samdark** [comment]
It seems tests are failing now a lot: https://travis-ci.org/yiisoft/yii2/builds/79668014

**Comment by SamMousa** [comment]
Yes they are not related to my code as far as I can tell.

They seem to be related to the mysql.SQL file.
For example null not allowed or testing if Null + 1 is 1; will look into
details tomorrow.

**Comment by samdark** [comment]
Not sure it's not related since previous commits are OK: https://travis-ci.org/yiisoft/yii2/builds/79646935

**Comment by dynasource** [comment]
Its related to this commit. I tested it locally: fails here too.

**Comment by dynasource** [comment]
The test issues are drilled down to BaseActiveRecord::getDirtyAttributes

**Comment by dynasource** [comment]
The fix for the PR of @SamMousa is to be found at #9660

**Comment by klimov-paul** [comment]
Duplicates #2262

**Comment by dynasource** [comment]
@klimov-paul: these PR's pretty much look like refactoring. If something is not alright, shouldnt tests be made to support this (and criteria mentioned in #2262)

**Comment by SamMousa** [comment]
Can you please stop instantly closing every issue / PR @klimov-paul ?

The attitude makes for a really hostile environment for contributors.
I'm taking the feedback I'm given and I've even posted yesterday that I'd look at the test details today; and you don't have the patience to leave this issue open for 24 hours...


-- I've updated the PR to pass all tests and have added a new test for getDirtyAttributes.

**Comment by samdark** [comment]
So now it seems there are no compatibility issues i.e. it's pure refactoring. That's great. Would be cool to test CPU performance and memory usage on a real world task with ~100 models.

**Comment by SamMousa** [comment]
Hmm, it shouldn't be too hard to do some basic testing with microtime (I'm not familiar with more advanced benchmarking tools). I'll post results later.

**Comment by samdark** [comment]
Great!

**Comment by SamMousa** [comment]
-- RESULTS --

Test script here: https://gist.github.com/SamMousa/30b2ad1d7931faf4398e

Tests are repeated 1000 times, with each run doing the get or set 500 times.

Executing Setting attributes (magic)...
Executing Setting attributes (setAttribute)...
Executing Getting attributes (magic)...
Executing Getting attributes (getAttribute)...
Results: (1000,500)
* master
 | Test                                     | Total    | Min      | Max      | Avg      | 
 | ---------------------------------------- | -------- | -------- | -------- | -------- | 
 | Setting attributes (magic)               | 3.9604   | 0.0032   | 0.0209   | 0.0040   | 
 | Setting attributes (setAttribute)        | 3.3561   | 0.0030   | 0.0048   | 0.0034   | 
 | Getting attributes (magic)               | 21.5356  | 0.0184   | 0.0378   | 0.0215   | 
 | Getting attributes (getAttribute)        | 2.0596   | 0.0018   | 0.0036   | 0.0021   | 
Results: (1000,500)
* fix-issue-9656
 | Test                                     | Total    | Min      | Max      | Avg      | 
 | ---------------------------------------- | -------- | -------- | -------- | -------- | 
 | Setting attributes (magic)               | 6.8812   | 0.0062   | 0.0318   | 0.0069   | 
 | Setting attributes (setAttribute)        | 3.4744   | 0.0030   | 0.0057   | 0.0035   | 
 | Getting attributes (magic)               | 19.8670  | 0.0180   | 0.0348   | 0.0199   | 
 | Getting attributes (getAttribute)        | 2.2408   | 0.0015   | 0.0157   | 0.0022   | 
Results: (1000,500)
* fix-issue-9656 -- with setAttribute improvement.
 | Test                                     | Total    | Min      | Max      | Avg      | 
 | ---------------------------------------- | -------- | -------- | -------- | -------- | 
 | Setting attributes (magic)               | 5.2172   | 0.0047   | 0.0306   | 0.0052   | 
 | Setting attributes (setAttribute)        | 3.6626   | 0.0032   | 0.0056   | 0.0037   | 
 | Getting attributes (magic)               | 23.2264  | 0.0181   | 0.1140   | 0.0232   | 
 | Getting attributes (getAttribute)        | 2.3164   | 0.0016   | 0.0063   | 0.0023   | 

This was run inside a Vagrant VM on a pretty old laptop.
Furthermore note that getAttribute shows a 15% increase in time, but the code hasn't changed (this indicates that there is a lot of noise in the test).

In both branches, i think __get can be improved a lot by caching the result of the attributes() function. Currently that function is used in hasAttribute which calls the function and then does in_array. If we cache the result and flip it we can use isset (which has constant time) instead of doing linear search.

In the fix branch setAttribute has an extra $check parameter to prevent hasAttribute from being called twice (once in __set and once in setAttribute).

**Comment by SamMousa** [comment]
Regardless of this PR, caching of attributes() increases speed of __get by about 1000% but it comes at a cost.
1. Increased memory usage. (minor)
2. If hasAttribute only gets called once for a class instance it will be slower. (minor)
3. The result of the attributes function is cached for the lifetime of the AR instance, this means overriding implementations should take that into account. (major: BC break in some instances)

**Comment by samdark** [comment]
So overall it's getting 6—15% slower for all types of access and during rendering a grid or a list of models attributes are accessed a lot. The pro is that there's a room for performance optimization.

**Comment by SamMousa** [comment]
Well, actually it will not be slower.
My test case (which I will amend maybe tomorrow or after the weekend) uses empty records for testing getAttribute. This is slow because the key is never set in $_attributes (forcing the in_array search in hasAttribute).

When rendering stuff it seems fair to assume most if not all of the fields are set (since we select * by default). 

So in practice I doubt there will be any noticeable differences for __get.
I'll try to adapt my benchmark to do some queries as well (fetching 1000 records and setting their attributes might be affected more than the getters)..

**Comment by samdark** [comment]
OK.

**Comment by SamMousa** [comment]
Okay, i've created a new branch that contains the optimizations. 
Also i've added some more test to the benchmark script and tried them on a VPS.

Results here: http://pastebin.com/gcy92f56

I have optimized the flow altering setAttribute.
Currently it returns void, now it returns true or false.
Currently it throws an exception, now you can pass a second argument to disable the exception.

Further optimization could be done by caching the attributes() function result, however:
- It needs to happen at a class level not object level since you are not likely to repeatedly (like 100.00+ get empty attributes for the same objects)

In conclusion I think the refactoring is okay and won't make a difference in real world applications. 

Regarding the optimizations, they change the signature of setAttribute, meaning everyone who is overrriding it will have to change their code.

Let me know what you decide or what more you'd like me to do!

**Comment by samdark** [comment]
Changing return value of `setAttribute` is a backwards compatibility break. It makes it impossible to merge changes into 2.0.x.

**Comment by SamMousa** [comment]
Yes I considered that; but since currently there is no return value I considered it would not be an issue.
The current return is void; so it makes no sense for any existing code to ever check the return value right?

**Comment by samdark** [comment]
Well, any change to public API is an issue. We're making exceptions for security matters but overall we'd like to keep codebase as backwards compatible as possible.

**Comment by samdark** [comment]
Is there an option to achieve it w/o changing public signature?

**Comment by SamMousa** [comment]
Yes ^^, basically like queryInternal we could do setAttributeInternal wich would return true / false and have setAttribute call that and return null / throw an exception.
It would not reduce performance much since internally we will use setAttributeInternal and skip the extra function call (which is not really an issue in any case)

**Comment by samdark** [comment]
Then probably it's good to do it compatible way.

**Comment by SamMousa** [comment]
Okay, please review the rest of my code as well so I can make all required changes in one go!

**Comment by samdark** [comment]
The rest looks OK to me.

**Comment by SamMousa** [comment]
New PR here: https://github.com/yiisoft/yii2/pull/9683
Feel free to close this one.
